### ISSUE TITLE
Invalid for_each argument on aws_default_tags

### ISSUE THREAD

**Issue by boopathykpm** [terraform-aws-modules/terraform-aws-eks#2281]
Getting the error while running the EKS cluster creation

│ Error: Invalid for_each argument
│
│   on .terraform/modules/eks/main.tf line 71, in resource "aws_ec2_tag" "cluster_primary_security_group":
│   71:   for_each = { for k, v in merge(var.tags, var.cluster_tags, data.aws_default_tags.current.tags) :
│   72:     k => v if local.create && k != "Name" && var.create_cluster_primary_security_group_tags
│   73:   }
│     ├────────────────
│     │ data.aws_default_tags.current.tags is a map of string, known only after apply
│     │ local.create is true
│     │ var.cluster_tags is map of string with 3 elements
│     │ var.create_cluster_primary_security_group_tags is true
│     │ var.tags is map of string with 3 elements

The "for_each" map includes keys derived from resource attributes that cannot be determined until apply, and so Terraform cannot determine the full set of keys that will identify the instances of this resource.

This is happening because of  `data "aws_default_tags" "current" {}`. 

There is no proper documentation to fix this error, any suggestions ?

**Comment by mmclane** [comment]
I am getting the same issue.  I have tried v18.30.0, v18.30.1, and v18.30.2.  Going back to v18.29.1 fixes the issue for me.

**Comment by earglass** [comment]
Hi! I have same issue but on different resource. What is more interesting, I'm getting this error on both eks and self managed node groups, for policy attachments, even though EKS group for me has only one policy which is hardcoded arn.

```
│ Error: Invalid for_each argument
│ 
│   on .terraform/modules/eks/modules/eks-managed-node-group/main.tf line 448, in resource "aws_iam_role_policy_attachment" "this":
│  448:   for_each = var.create && var.create_iam_role ? toset(compact(distinct(concat([
│  449:     "${local.iam_role_policy_prefix}/AmazonEKSWorkerNodePolicy",
│  450:     "${local.iam_role_policy_prefix}/AmazonEC2ContainerRegistryReadOnly",
│  451:     var.iam_role_attach_cni_policy ? local.cni_policy : "",
│  452:   ], var.iam_role_additional_policies)))) : toset([])
│     ├────────────────
│     │ local.cni_policy is a string, known only after apply
│     │ local.iam_role_policy_prefix is a string, known only after apply
│     │ var.create is true
│     │ var.create_iam_role is true
│     │ var.iam_role_additional_policies is list of string with 1 element
│     │ var.iam_role_attach_cni_policy is true
│ 
│ The "for_each" set includes values derived from resource attributes that
│ cannot be determined until apply, and so Terraform cannot determine the
│ full set of keys that will identify the instances of this resource.
│ 
│ When working with unknown values in for_each, it's better to use a map
│ value where the keys are defined statically in your configuration and where
│ only the values contain apply-time results.
│ 
│ Alternatively, you could use the -target planning option to first apply
│ only the resources that the for_each value depends on, and then apply a
│ second time to fully converge.
```
`iam_role_additional_policies = ["arn:aws:iam::aws:policy/CloudWatchAgentServerPolicy"]`

**Comment by bryantbiggs** [comment]
the use of default tags added in v18.30 was unfortunately found to have introduced some issues and we will be correcting this in v19 since its removal is a breaking change. Apologies for any headache experienced by this

**Comment by antonbabenko** [comment]
This issue has been resolved in version 19.0.0 :tada:
